lasagne/layers/pool.py

```diff
@@ -15,81 +14,212 @@
 ]
 
 
-def pool_output_length(input_length, pool_length, ignore_border=True):
-    '''Compute the output length of a pooling operator
-    along a particular dimension.
+def pool_output_length(input_length, ds, st, ignore_border=True, pad=0):
+    """
+    Compute the output length of a pooling operator
+    along a single dimension.
 
     Parameters
     ----------
-    input_length
-    pool_length
-        Shape of the input and pooling operator in the chosen dimension
-
-    ignore_border:
-        if True, the output length is rounded down.
-        if False, it is rounded up
+    input_length : integer
+        The length of the input in the pooling dimension
+    ds : integer
+        The length of the pooling region
+    st : integer
+        The stride between successive pooling regions
+    pad : integer
+        The number of elements to be added to the input on each side.
+    ignore_border: bool
+        If True, partial pooling regions will be ignored.
 
     Returns
     -------
     output_length
-        * None if either input is None
-        * `input_length / pool_length` otherwise
-    '''
-
-    if input_length is None or pool_length is None:
+        * None if either input is None.
+        * Computed length of the pooling operator otherwise.
+
+    Notes
+    -----
+    When ignore_border = True, this is given by the number of full
+    pooling regions that fit in the padded input length,
+    divided by the stride (rounding down).
+
+    When ignore_border = False, partial pooling regions are also
+    allowed, as long as each pooling region contains at least one element
+    not present in another pooling region.
+    """
+    if input_length is None or ds is None:
         return None
 
     if ignore_border:
-        # Round shape down when we ignore the border
-        return input_length // pool_length
+        output_length = input_length + 2 * pad - ds + 1
+        output_length = (output_length + st - 1) // st
+
+    # output length calculation taken from:
+    # https://github.com/Theano/Theano/blob/master/theano/tensor/signal/downsample.py
+    else:
+        if st >= ds:
+            output_length = (input_length + st - 1) // st
+        else:
+            output_length = max(
+                0, (input_length - ds + st - 1) // st) + 1
```

Comment:
I think you're missing the padding here. [In theano.tensor.signal.downsample](https://github.com/Theano/Theano/blob/master/theano/tensor/signal/downsample.py#L143-144), they pad unconditionally before checking for `ignore_border`. If I'm right, there should be a test for this scenario, otherwise a comment explaining it would be great.